XiEditor/StyleMap.swift

```diff
@@ -16,11 +16,55 @@ import Cocoa
 
 /// A represents a given text style.
 struct Style {
-    var fgColor: NSColor;
-    var bgColor: NSColor;
-    var weight: UInt16;
-    var underline: Bool;
-    var italic: Bool;
+    var font: NSFont?
+    var fgColor: NSColor?
+    var bgColor: NSColor?
+    var underline: Bool
+    var italic: Bool
+    var weight: Int?
+    var attributes: [String: Any] = [:]
+
+    init(font fromFont: NSFont, fgColor: NSColor?, bgColor: NSColor?, underline: Bool, italic: Bool, weight: Int?) {
+        if let fgColor = fgColor {
+            attributes[NSForegroundColorAttributeName] = fgColor
+        }
+
+        if let bgColor = bgColor, bgColor.alphaComponent != 0.0 {
+            attributes[NSBackgroundColorAttributeName] = bgColor
+        }
+
+        if underline {
+            attributes[NSUnderlineStyleAttributeName] = NSUnderlineStyle.styleSingle.rawValue
+        }
+
+        let fm = NSFontManager.shared()
+        var font: NSFont?
+
+        if italic {
+            var traits = fm.traits(of: fromFont)
+            traits.insert(NSFontTraitMask.italicFontMask)
+            if let f = closestMatch(of: fromFont, traits: traits, weight: weight ?? fm.weight(of: fromFont)) {
+                font = f
+            } else {
+                attributes[NSObliquenessAttributeName] = 0.2
+            }
+        }
+
+        if font == nil && weight != nil {
+            font = closestMatch(of: fromFont, traits: fm.traits(of: fromFont), weight: weight!) ?? fromFont
```

Comment:
Actually, the `?? fromFont` wasn't supposed to be there anymore, since I prefer to not set the `NSFontAttributeName` if the default font is used. I'll remove it